fix(FindUs): Water matches event's theme

retrohacker opened the pull request:
<!-- 👋 Hi, thanks for sending a PR to halfstackconf-next! 💖.
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR. -->

## PR Checklist

- [x] Addresses an existing open issue: fixes #16 
- [x] That issue was marked as [accepting prs](https://github.com/JoshuaKGoldberg/halfstackconf-next/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/halfstackconf-next/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

FindUs' Google Map embed's water color now matches the event's theme

Attempted getComputedStyles on a throw-away div created with document.createElement, but didn't have any luck and it broke SSR:

```
const div = document.CreateElement("div")
div.classList.add(styles[event])
document.getComputedStyles(div)
```

Instead, this PR pulls those variables into JS. Not sure if there is a better way to do this.

retrohacker committed:
fix(FindUs): Water matches event's theme

FindUs' Google Map embed's water color now matches the event's
theme.

Attempted getComputedStyles on a throw-away div created with
document.createElement, but didn't have any luck and it broke SSR:

  const div = document.CreateElement("div")
  div.classList.add(styles[event])
  document.getComputedStyles(div)

Instead, this PR pulls those variables into JS.

JoshuaKGoldberg commented on src/components/EventTheme/colors.tsx:
```diff
@@ -0,0 +1,67 @@
+    "color-primary-dark": "#304007",
+    "color-primary-light": "#e1ff99",
+  },
+} as Record<
```
[Suggestion] Heh, the `as` type assertion is because keys of type `string` can't be used to index into this object, right? Until #18 is in I suppose there will always need to be some assertions. A fewer-lines-of-code solution would be to, instead of `as` asserting here, using an `as` in `EventColors`:

```ts
return colors[event as keyof typeof colors];
```

That way also has the added benefit of letting `colors` keep its more specific type. Which might be useful later on.


<details>
<summary>If we really wanted to, we could go through the effort of making it more safe...</summary>
I think TypeScript is right to yell here: what happens if someone makes a new event -say, `"philadelphia"`- and forgets to add to `colors.tsx`? The page will crash. I simulated this by deleting the `charlotte` entry and visiting `localhost:3000/charlotte` locally:

```plaintext
TypeError: Cannot read properties of undefined (reading 'color-backdrop-subtle')

src/pages/[event]/index.tsx (124:34) @ Event
  122 |     geolocation={geolocation}
  123 |     slug={slug}
> 124 |     waterColor={eventColors["color-backdrop-subtle"]}
      |                            ^
  125 |   />
  126 | )}
```

An alternative might be to instead change from `event: string` to a more specific type, like `event: EventName` or `event: EventSlug` or some name like that.

One way you could do that would be to:
1. Change `src/data/events.json` to `src/data/events.ts`
2. Have it `export const events = [...] as const` _(maybe even renaming it to `eventsOrder`, since that's what it's used for?)_
3. Also have `events.ts` `export const EventName = typeof events[number]`
4. Use that `EventName` type wherever something is meant to be an event

Note that this includes `getEvents()` in `src/data/index.ts`, which would need to `as EventSlug[]`... which I guess means there's always going to be be a type assertion until #18 is in.
</details>

retrohacker replied:
Love this, going to give it a try. I think it'll end up being a bit bigger PR though.

Learned a lot of TS on this change so far. Didn't know you could do this with `as const`.

JoshuaKGoldberg commented on src/components/EventTheme/index.tsx:
```diff
@@ -1,14 +1,23 @@
   );
 }
+
+export function EventColors({ event }: EventColorProps) {
+  return colors[event];
+}
```
[Suggestion] This component and its prop type are named & structured like a React function component, but it's not actually used like one. It's used more like a _hook_. I think structuring it like one would be more idiomatic (and less code):

```suggestion
export function useEventColors(event: string) {
  return colors[event];
}
```

Thoughts?

JoshuaKGoldberg commented on src/pages/[event]/index.tsx:
```diff
@@ -36,6 +36,7 @@ export default function Event({
     year,
   },
 }: ReturnedProps<typeof getStaticProps>) {
+  const eventColors = EventColors({ event: slug });
```
[Suggestion] I would move this into `FindUs`. The `waterColor` value isn't used until the `onLoad` callback inside that component, and the component also receives the `slug`. Fewer lines of code for the same result.

retrohacker replied:
Good call!

JoshuaKGoldberg commented on src/components/FindUs/createMapOptions.ts:
```diff
@@ -73,9 +77,7 @@ export const createMapOptions = () => ({
-        {
-          color: "#072a74",
-        },
+        { color: waterColor },
```
[Praise] Ah nice, glad it was this clean!

JoshuaKGoldberg reviewed (changes requested):
Cool, nicely done getting this to work! It's been bugging me for a while 😄 

Left a couple of code suggestions. What do you think?

retrohacker committed:
refactor(EventName): typescript is now aware of event names

./src/data/events/index.ts now exports the event names, in order,
as a const assertion.

We then use that in ./src/data/types.ts to generate the EventName
type.

We then update all instances of `event: string` and `slug: string`
to use the narrower EventName instead.

This commit also adds type checking to the shape of objects in the
colors array inside src/components/EventTheme/colors.tsx

This PR leaves the Tito-Widget event name as a string since it
appears to support multiple event name encodings.

retrohacker commented:
Thanks for the kind review and guidance, learned a lot about TS with this PR.

Let me know what you think!

JoshuaKGoldberg commented on src/data/types.ts:
```diff
@@ -80,10 +82,12 @@ export interface EventVideo {
+  slug: EventName;
 }
+
+export type EventName = (typeof eventOrder)[number];
```
[Suggestion] Looking through this PR, there are 15 `: EventName`s, 8 `event: EventNames`s, and 7 `slug: EventName`s. Maybe `EventSlug` would be more clear?

Not a blocker from me, just a thought 🙂

JoshuaKGoldberg commented on README.md:
```diff
@@ -114,7 +114,7 @@ pm2 start "pnpm production"
 Events are all stored in `src/data` as `.json` files.
 
-- [`src/data/events.json`](./src/data/events.json) contains the array of event lines that are listed on the homepage
+- [`src/data/events/index.ts`](./src/data/events/index.ts) contains the array of event lines that are listed on the homepage
```
[Praise] I probably would have forgotten to update these docs. Nice! 😄

JoshuaKGoldberg commented on src/components/EventTheme/colors.tsx:
```diff
@@ -0,0 +1,76 @@
+import { EventName } from "~/data/types";
+
+interface EventColors {
```
[Suggestion] Since this interface is used in things available outside of this file (the type of `colors`), you might want to export it. It doesn't make a difference now but later on someone else might want to refer to it, and it can be confusing not having the interface available.

```suggestion
export interface EventColors {
```

(non-blocking, just a thought)

JoshuaKGoldberg commented on src/components/EventTheme/colors.tsx:
```diff
@@ -0,0 +1,76 @@
+    "color-primary-light": "#e1ff99",
+  },
+};
+export default colors;
```
[Style] I personally tend to avoid `export default`: https://basarat.gitbook.io/typescript/main-1/defaultisbad. Definitely not a blocker, but I'd suggest switching to `export const colors`.

JoshuaKGoldberg commented on src/components/FindUs/createMapOptions.ts:
```diff
@@ -1,4 +1,8 @@
-export const createMapOptions = () => ({
+interface createMapArguments {
```
[Style] It's generally convention to PascalCase type names, not camelCase.

```suggestion
interface CreateMapArguments {
```

JoshuaKGoldberg reviewed (approved):
Awesome! Really thorough job hitting all the places. Even the docs! 👏 

Left a few suggestions/thoughts but nothing blocking. Yay!

JoshuaKGoldberg commented:
By the way,

> ## PR Checklist
> * [ x ] Addresses an existing open issue: fixes #000
> * [ x ] That issue was marked as [accepting prs](https://github.com/JoshuaKGoldberg/halfstackconf-next/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22+sort%3Aupdated-desc)
> * [ x ] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/halfstackconf-next/blob/main/.github/CONTRIBUTING.md?rgh-link-date=2023-09-12T18%3A59%3A12Z) were taken

f you switch the `[ x ]`s to `[x]` GitHub will render as a nice task list: https://docs.github.com/en/issues/managing-your-tasks-with-tasklists/creating-a-tasklist

> fixes #000

`fixes #16`?

> Instead, this PR pulls those variables into JS. Not sure if there is a better way to do this.

Yeah 🤷 if the projects eventually moves to CSS-in-JS (#17) this'll all get resolved.

retrohacker committed:
refactor(EventTheme): Code style tweaks

Exports EventColors from EventTheme

Avoids export default

PascalCase typescript names

EventName -> EventSlug
